### DISCUSSION TITLE
How to bind huge classes

### DISCUSSION THREAD

**Discussion by ghorn** [wjakob/nanobind#759] [Q&A] [answered]
I embarrassingly ran into the 1<<24 byte size limit on a class and triggered the static_assert. I couldn't find a direct reference to this limit in the docs, did I miss it?

Is nanobind designed to support objects that must be entirely allocated on the heap? I was actually surprised to see the limit, since I thought everything would be done on the heap anyway. How do I bind this sort of struct?

**Comment by ghorn** [comment]
I mean the obvious thing to do would be to
```c++
struct Wrapper {
   unique_ptr<UnreasonablyBigStruct> ptr;
};
```
Just wondering if there's a different way to do it.

**Reply by ghorn** [reply]
The wrapper approach breaks down when I do something like `bind_vector.h`. I'm actually not sure what to do at this point.

Where does this size limit come from?

**Comment by ghorn** [comment]
I tried simply patching out the `static_assert` and things seem to work.... I have a pretty good test suite that is passing with a 1GB class. I found all the accidental copies and stack allocation I had in user code and converted it to not copy / allocate on the heap.

It would be very helpful to get a comment on this from the author/maintainer, to know if there are are any landmines here.

**Reply by wjakob** [reply, marked as answer]
The static_assert is needed. The information that nanobind stores about the type is fairly compressed, and only 24 bits are available to encode the size. You will have undefined behavior if nanobind ever tries to allocate such a class on its own (instead of allocating one in C++ and then returning it to Python), as it will not acquire enough memory for it.

**Reply by ghorn** [reply]
Thank you. Sounds like `Wrapper` struct is the only sound approach. When does nanobind allocate a class on its own? Is it feasible for me to simply avoid such operations?

**Reply by wjakob** [reply]
When you call the `__init__` constructor from Python. If the only way to create the class is by return value, then the Python allocation path will not be used.

**Reply by wjakob** [reply]
Also, return value policies that copy will have an issue.

**Reply by ghorn** [reply]
Thanks for the information. I'm not confident I'd get this right to go through with the unsafe version.

One long term idea for nanobind to support large classes would be to move the `static_assert` to the places where that size is accessed. The error message would be `"classes above 16MB can't be returned this way, you must allocate them explicitly with new"` or similar.

**Reply by ghorn** [reply]
Can you actually elaborate?

> When you call the `__init__` constructor from Python. If the only way to create the class is by return value, then the Python allocation path will not be used.

I don't quite get this. The only patterns i'm using are "built-in" constructors like
```c++
    .def(nb::init<>())
```
and the manual ones like:
```c++
    // by pointer
    .def(
      "__init__",
      [](Foo* self) {
             new (self) Foo();
      });
    // by reference
    .def(
      "__init__",
      [](Foo& self) {
             new (&self) Foo();
      });
```
do those use the "safe" path?

**Comment by ghorn** [comment]
I'm reading the code and I see the limitation on number of flags (24 bits). But to me, it looks like max size is a full 32 bits:

https://github.com/wjakob/nanobind/blob/bbbd022229bf466073288a913a640172fc4a7842/include/nanobind/nb_class.h#L94-L98

am I misunderstanding? Or is the size limited in a different struct?

**Reply by wojdyr** [reply]
This static_assert was added when the size was changed to 24 bytes in 43e179c09de67b57575e5e29612586ea9ba792f7 and then hasn't been removed when the size was changed back to 32 bits in
https://github.com/wjakob/nanobind/commit/0b3548ad6faf4eb18aa72259c3b82ac6df284652#diff-2a6e239c3e33400e1251b4087a19af93b4fd4886009de7e1255968681cf300e6L63-L67

**Reply by ghorn** [reply]
huzzah!

**Reply by ghorn** [reply]
Would you accept a PR changing the limit to `1 << 32`?

**Comment by wjakob** [comment]
Fixed in 2aacdcfdd312dabda3d65e683e418e4b3c8c0b04

**Reply by wojdyr** [reply]
it won't work with 32-bit size_t, but I don't know if any 32-bit platforms are supported

**Reply by wjakob** [reply]
Good catch!
